engines/wintermute/base/gfx/opengl/meshx.cpp

```diff
@@ -0,0 +1,498 @@
+		if (_materials[i]->getName() && _materials[i]->getName() == matName) {
+			_materials[i]->setSprite(sprite);
+		}
+	}
+	return true;
+}
+
+//////////////////////////////////////////////////////////////////////////
+bool MeshX::setMaterialTheora(const Common::String &matName, VideoTheoraPlayer *theora) {
+	for (uint32 i = 0; i < _materials.size(); i++) {
+		if (_materials[i]->getName() && _materials[i]->getName() == matName) {
+			_materials[i]->setTheora(theora);
+		}
+	}
+	return true;
+}
+
+//////////////////////////////////////////////////////////////////////////
+bool MeshX::invalidateDeviceObjects() {
+	// release buffers here
+
+	for (uint32 i = 0; i < _materials.size(); i++) {
+		_materials[i]->invalidateDeviceObjects();
+	}
+
+	return true;
+}
+
+//////////////////////////////////////////////////////////////////////////
+bool MeshX::restoreDeviceObjects() {
+	for (uint32 i = 0; i < _materials.size(); i++) {
+		_materials[i]->restoreDeviceObjects();
+	}
+
+	if (_skinnedMesh) {
+		return generateMesh();
+	} else {
+		return true;
+	}
+}
+
+bool MeshX::parsePositionCoords(XFileLexer &lexer, float *vertexCoords) {
+	for (uint i = 0; i < _vertexCount; ++i) {
+		for (int j = 0; j < 3; ++j) {
+			_vertexPositionData[i * 3 + j] = readFloat(lexer);
+			vertexCoords[i * 8 + 5 + j] = _vertexPositionData[i * 3 + j];
+		}
+
+		_vertexPositionData[i * 3 + 2] *= -1.0f;
+		vertexCoords[i * 8 + 5 + 2] *= -1.0f;
+
+		lexer.advanceToNextToken(); // skip semicolon
+	}
+
+	return true;
+}
+
+bool MeshX::parseFaces(XFileLexer &lexer, int faceCount) {
+	for (int i = 0; i < faceCount; ++i) {
+		int indexCount = readInt(lexer);
+
+		// we can add something to triangulize faces later if the need arises
+		if (indexCount != 3) {
+			warning("MeshX::loadFromX non triangle faces are not supported yet");
+			return false;
+		}
+
+		for (int j = 0; j < 3; ++j) {
+			_indexData[i * 3 + j] = readInt(lexer);
+		}
+
+		uint16 tmp = _indexData[i * 3 + 0];
+		_indexData[i * 3 + 0] = _indexData[i * 3 + 2];
+		_indexData[i * 3 + 2] =  tmp;
+
+		lexer.advanceToNextToken(); // skip semicolon
+	}
+
+	return true;
+}
+
+bool MeshX::parseTextureCoords(XFileLexer &lexer, float *vertexCoords) {
+	// should be the same as _vertexCount
+	int textureCoordCount = readInt(lexer);
+
+	for (int i = 0; i < textureCoordCount; ++i) {
+		vertexCoords[i * 8 + 0] = readFloat(lexer);
+		vertexCoords[i * 8 + 1] = readFloat(lexer);
+		lexer.advanceToNextToken(); // skip semicolon
+	}
+
+	if (lexer.reachedClosedBraces()) {
+		lexer.advanceToNextToken();
+	} else {
+		warning("Missing } in mesh object");
+	}
+
+	return true;
+}
+
+bool MeshX::parseNormalCoords(XFileLexer &lexer, float *vertexCoords) {
+	// should be the same as _vertex count
+	int vertexNormalCount = readInt(lexer);
+
+	for (int i = 0; i < vertexNormalCount; ++i) {
+		vertexCoords[i * 8 + 2] = readFloat(lexer);
+		vertexCoords[i * 8 + 3] = readFloat(lexer);
+		vertexCoords[i * 8 + 4] = -readFloat(lexer); // mirror

```

Comment:
mirror because?

Reply:
Shouldn't the normal vectors also mirrored at the x-y plane during loading, just like the position vectors? Or do you mean adding more context to the comment?